subsys/bluetooth/host/hci_core.h

```diff
@@ -244,6 +244,24 @@ struct bt_dev {
 #else
 	/* Pointer to reserved advertising set */
 	struct bt_le_ext_adv    *adv;
+#if (CONFIG_BT_ID_MAX > 1) && (CONFIG_BT_EXT_ADV_MAX_ADV_SET > 1)
+	/* When supporting multiple concurrent connectable advertising sets
+	 * with multiple identities, we need to know the identity of
+	 * the terminating advertising set to identify the connection object.
+	 * The identity of the advertising set is determined by its
+	 * advertising handle, which is part of the
+	 * LE Set Advertising Set Terminated event which is always sent
+	 * _after_ the LE Enhanced Connection complete event.
+	 * Therefore we need cache this event until its identity is known.
+	 *
+	 * When an advertiser times out, it is only the high duty cycle
+	 * advertiser that generates the LE Enhanced Connection Complete event.
+	 * Therefore we use a state variable to determine if the cached event
+	 * should be processed or not.
+	 */
+	bool conn_complete_received;
+	struct bt_hci_evt_le_enh_conn_complete cached_conn_complete;
```

Comment:
I believe this can lead to a race condition if two advertisers manage to establish a connection before the TERMINATED is received for the first one.

Should this scale by CONFIG_BT_ID_MAX?

Reply:
I'm assuming that the race condition you are considering is in step 5 and 6 below:
1. Start advertising set 0 using identity 0
2. Start advertising set 1 using identity 1
3. Receive LE Connection Complete for advertising set 0
4. Receive LE Connection Complete for advertising set 1
5. Receive LE Advertising Set terminated event for advertising set 0
6. Receive LE Advertising Set terminated event for advertising set 1

If that sequence occurs, the host has no way of mapping the LE Advertising Set Terminated event to the Connection Complete event. That is, it can't map an identity to the newly established connection. 

Therefore, according to errata 14871, the controller should guarantee that the Advertising Set Terminated event follows the Connection Complete event. Hence, the controller should prevent sequences like the one above from happening.

Reply:
what about the warning in hci_core:1058? will that be raised in such scenario?

Reply:
Yes, this warning (https://github.com/zephyrproject-rtos/zephyr/pull/34848/files#diff-774b3de86adb275da54becf2a0958a2d4dae03d63e47840037ecc575e9d658b4R1058 ) will be printed in case the controller is raising another enhanced connection complete event before the LE Advertising Set Terminated event is processed

Reply:
The idea of caching the connection complete event is sound I believe. Since the terminated event contains a connection handle, we can do a lookup from the adv set to the enhanced connection complete event. 

The current implementation would not support the scenario above, where we receive 2 connection complete events before the advertising set terminated event, it only supports caching of one such even. 

I think this could be implemented as

1. Only Enchance Connection Complete Event, store the event in an array (same size as `MIN(CONFIG_BT_MAX_CONN, CONFIG_BT_EXT_ADV_MAX_ADV_SET)`)
2. On Advertising Set terminated lookup the connection handle to find the matching enhanced connection complete event 
3. Finalize connection complete

Furthermore, I believe this PR is missing some crucial connection cancellation features, e.g. if the application stops the advertising set before advertising set terminated event, but after the connection complete event (corner case, but can still cause a desync).

Reply:
@Thalley , thanks for your review. I believe I've addressed your comments by doing the following:
* Checking if the advertiser is still running before processing the cached connection complete event to avoid that the application gets a connection complete callback after stopping the advertiser.
* No longer depend on the order of events from the controller. Now the host caches all events that result in a connection being established. The host will compare the connection handles in those two events in order to create the mapping. The cost is some extra RAM usage.